lib/salus/scanners/npm_audit.rb

```diff
@@ -1,68 +1,287 @@
-          exceptions&.each do |exception|
-            if exception.keys.sort != %w[advisory_id changed_by notes]
-              report_error("The exception #{exception} doesn't have a proper format! Please ensure"\
-              ' that each exception has an `advisory_id`, `changed_by`, and `notes` field!')
+        ensure_package_lock! do
+          raw = run_shell(AUDIT_COMMAND).fetch(:stdout)
+          json = JSON.parse(raw, symbolize_names: true)
+
+          if json.key?(:error)
+            code = json[:error][:code] || '<none>'
+            summary = json[:error][:summary] || '<none>'
+
+            message =
+              "`#{AUDIT_COMMAND}` failed unexpectedly (error code #{code}):\n" \
+              "```\n#{summary}\n```"
+
+            log(colorize(message, :red))
+            return false
+          end
+
+          raw_advisories = json.fetch(:advisories).values
+
+          advisories = raw_advisories.map do |advisory|
+            # findings is an array of objects looking roughly like
+            # [
+            #   { "version": "1.0.5", ..., "dev": false },
+            #   { "version": "1.0.5", ..., "dev": true }
+            # ]
+            # where each element records an instance of the vulnerable package
+            # in the dependency tree. If, for some finding, dev is false,
+            # then there exists a vulnerable version of the module in the prod
+            # dependency tree
+            prod = advisory.fetch(:findings).any? { |finding| !finding.fetch(:dev) }
+
+            id = advisory.fetch(:id).to_s
+            excepted = @exceptions.include?(id)
+
+            Advisory.new(
+              id,
+              advisory.fetch(:module_name),
+              advisory.fetch(:title),
+              advisory.fetch(:severity),
+              advisory.fetch(:url),
+              prod,
+              excepted
+            )
+          end
+
+          # Sort advisories with un-excepted prod vulnerabilities first,
+          # and in ascending order of ID otherwise
+          advisories.sort_by! do |advisory|
+            unexcepted_and_prod = !advisory.excepted? && advisory.prod?
+            [(unexcepted_and_prod ? 0 : 1), advisory.id.to_i]
+          end
+
+          advisories_by_id = advisories.map(&:id).zip(advisories).to_h
+
+          # The build fails iff there exist un-expect advisories against prod deps
+          failing_advisory_ids = advisories.select { |adv| !adv.excepted? && adv.prod? }.map(&:id)
+
+          # For informational purposes, we record any exceptions that don't refer
+          # to an actual advisory
+          extraneous_exceptions = @exceptions - advisories_by_id.keys
+
+          # Also, we record any exceptions that apply only to dev dependencies
+          extraneous_dev_exceptions = @exceptions.select do |id|
+            adv = advisories_by_id[id]
+            !adv.nil? && !adv.prod?
+          end
+
+          if advisories.any?
+            log(tabulate_advisories(advisories))
```

Comment:
need to put `report_failure` above this line.
I'd then loop through and log each advisory similar to how we do here: https://github.com/coinbase/salus/blob/master/lib/salus/scanners/bundle_audit.rb#L74

I'll write a comment on the PR in a bit regarding the `log()` code.